### ISSUE TITLE
TypeError in WebSocketServer.wait_closed() since tasks not initialized

### ISSUE THREAD

**Issue by cjerdonek** [aaugustin/websockets#309]
If an exception occurs in `WebSocketServerProtocol.process_request()` and the server is closed, `WebSocketServer.wait_closed()` can raise an exception like the following:

      File "/.../websockets/server.py", line 582, in wait_closed
        loop=self.loop)
      File "/usr/local/lib/python3.6/asyncio/tasks.py", line 311, in wait
        fs = {ensure_future(f, loop=loop) for f in set(fs)}
      File "/usr/local/lib/python3.6/asyncio/tasks.py", line 311, in <setcomp>
        fs = {ensure_future(f, loop=loop) for f in set(fs)}
      File "/usr/local/lib/python3.6/asyncio/tasks.py", line 526, in ensure_future
        raise TypeError('An asyncio.Future, a coroutine or an awaitable is '
    TypeError: An asyncio.Future, a coroutine or an awaitable is required

This looks to be because the tasks that the server is waiting on in `wait_closed()` are not yet initialized (since the initialization was interrupted by the exception in `process_request()`):

https://github.com/aaugustin/websockets/blob/136d9bc9a917f7f5fdde4001e801b591ee1b4a0b/websockets/server.py#L578

**Comment by cjerdonek** [comment]
Looking at the code, it looks like it's the `close_connection_task` which is at fault because it isn't initialized until `connection_open()`, which happens at the end of `handshake()` (and in particular after `process_request()`).

**Comment by cjerdonek** [comment]
Actually, it's possible this is an issue also in the case that `process_request()` raises no exception but returns a non-None value. I'll have to look into it more.

**Comment by cjerdonek** [comment]
[This discussion](https://github.com/aaugustin/websockets/pull/265/files#r140941685) on the topic of whether there should be a separate `close_connection_task` (at the time it was introduced) is also relevant.

It seems like there are many possible variations / ways of fixing this issue, but it's hard for me to evaluate which is best without developing a better understanding of the motivation for having `close_connection_task`. One easy way to fix it might be to initialize the attribute with a "no-op" coroutine that can be awaited on at any point: `noop()`, where `async def noop(): pass`, or maybe there is an even simpler way to express such a thing. Another would be to introduce another boolean flag to store how far things have progressed -- an approach I would lean against. Yet another would be simply to check whether the task has been set before trying to wait on it. The first seems the most elegant because it doesn't introduce additional if statements. (I still have an instinctive feeling that there should be a way to remove `close_connection_task` entirely while also simplifying reasoning about the logic.)

**Comment by cjerdonek** [comment]
I posted PR #310 to fix this.

A couple notes on the fix:

1. On the issue itself, I realized that `WebSocketServer.wait_closed()` shouldn't be waiting on `close_connection_task` in the first place. i believe `close_connection_task` should be an implementation detail that higher-level constructs like `WebSocketServer` shouldn't have to know about to manage the connection. This is supported by `handler_task`'s docstring which says, "Handle the lifecycle of a WebSocket connection." If there is a case where waiting only on `handler_task` isn't sufficient, I think that probably indicates a problem with the logic in `handler` and should be fixed.

2. Also, the test I added is an "async" test (decorated with `@asyncio.coroutine`). Sometime in the past I noticed that the tests could be simplified and made cleaner by converting many of the tests to this form. The main advantage is that the tests are simpler because you don't have to interact with the loop at all in the body of the test. (The loop is handled by the decorator, in the same way that a script's `main()` function would handle creating the loop.) This makes the tests "pure" async. In this case, I was having trouble creating a failing test with the existing "sync" test approach, so I used this as an opportunity to introduce the first async test. If more tests are made pure async, the code can be made more DRY with the other tests, as well as creating helper functions for starting and stopping the server and connections, etc.

**Comment by cjerdonek** [comment]
Something else I noticed while working on and thinking about this issue is that it seems like it might be simpler and better if `WebSocketCommonProtocol.close()` is a non-blocking (i.e. non-async) function that simply tells the connection / handler task to start its close logic. This would also parallel the API for how [`WebSocketServer.close()`](https://github.com/aaugustin/websockets/blob/136d9bc9a917f7f5fdde4001e801b591ee1b4a0b/websockets/server.py#L535) and asyncio's [`Server.close()`](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server.close) work. Currently, it doesn't seem like there is a way just to tell a connection to close. You have to wait on the `close()` coroutine which seems redundant because `handler()` already has logic to wait on `close()`. This might be one of the things contributing to making the logic harder to reason about as there is more than one function responsible for the same thing.

**Comment by aaugustin** [comment]
Indeed I didn't update `wait_closed` correctly when I switched to the two-task design.

To recap, here's the high level flow on the server side:

- when the TCP connection is open, `connection_made`:
    - registers it in `self.ws_server.websockets` — whose only purpose is to keep track of connections in order to shut them down properly when closing the server
    - starts `handler_task` which executes `handler`, which contains three blocks

- the first block of `handler` executes the opening handshake
    - if `handshake` succeeds, it calls `connection_open` which starts `transfer_data_task` and `close_connection_task`; both tasks are started at the same time so the one task vs. two task design question doesn't matter here ; note that we have three tasks at this point: `handler_task`, `transfer_data_task` and `close_connection_task`.
    - if `handshake` exits prematurely (because `process_request` aborted the handshake) or fails with an exception, `connection_open` is never called; `transfer_data_task` and `close_connection_task` aren't initialized

----

`wait_closed` must guarantee that all TCP connections are closed before it returns.

When `handshake` succeeded, `close_connection_task` takes care of that. Waiting on `close_connection_task` in `wait_closed` is correct.

The case when `handshake` exited prematurely isn't handled correctly. To the best of my understanding, that's the problem you report.

Error handling in that case isn't great:

```
        except Exception:
            # Last-ditch attempt to avoid leaking connections on errors.
            try:
                self.writer.close()
            except Exception:                               # pragma: no cover
                pass
```

Since then I understood a lot better how asyncio APIs map to TCP concepts — with the additional complexity added by TLS — and that's why `close_connection` is much more elaborate.

So the question becomes — **how does `wait_closed` ensure a connection is closed when it doesn't go through through `connection_made` -> start `handler_task` -> `handshake` -> `connection_open` -> start `connection_close_task` code path?**

Perhaps `wait_closed` should start a `connection_close_task` for connections that don't have one running?

----

Regarding your last comment, user code usually won't call `close` on the server side. It will merely exit the handler and let `websockets` deal with the closing handshake.

Users shouldn't even think about the websocket opening and closing handshakes. (Of course, we do so they don't have to!)

**Comment by cjerdonek** [comment]
Thank you, @aaugustin! :) I really appreciate you giving this the thought you did in your reply.

I've been mulling over this issue myself on and off over the past several days and can share some of my thoughts when they're a bit clearer and more focused.

In the meantime, I wanted to respond briefly to this last comment (and then I will read and reflect more on what you have written, too):

> Regarding your last comment, user code usually won't call close on the server side.

This is actually the case that I've been running into, so I'll explain the use case so you can see how it has been arising.

I have a connection handler on the server-side that fires off a number of tasks associated with the connection over the course of the connection's lifetime -- using `asyncio.ensure_future()`. If one of those tasks encounters an exception, I want the task to be able to close the associated connection (without having to block or create yet another task), which should in turn cause the cancellation / clean-up of the other tasks. When the task initiates the close, this should cause the parent handler to clean-up the tasks it spawned.

When I studied the websockets code base to see how it might handle any similar situation, I found that the server does this when it calls `websocket.handler_task.cancel()` on each connected client when it is cleaning up:

https://github.com/aaugustin/websockets/blob/136d9bc9a917f7f5fdde4001e801b591ee1b4a0b/websockets/server.py#L559

This is basically the API I was looking for because it is non-blocking, and because the `CancelledError` will bubble up to the parent handler and cause the clean-up as I wanted. The main downside though is that this isn't part of the public API, and it's not entirely clear if it has the guarantees I want (though it probably does given that it's used by the server).

It seems like it would be nice if, instead of having to separately wait on a `close()` coroutine, users could just call a non-blocking `close()` function and then wait on the connection as a whole. (This is a bit like the `close()` / `wait_closed()` API we see in `asyncio`.) One of the sticky points of the current API that we both know about is that by exposing multiple tasks / coroutines to the end-user, you have to worry about how to handle the end-user trying to `cancel()` each of these. If there is only one "root" task that is exposed to the end-user (namely the connection itself), the API becomes much simpler to explain. For example, canceling it could be documented as an idempotent operation that causes the close sequence to initiate (e.g. proceeding to the finally block of a try-finally, which could be shielded), and `close()` could be implemented as calling `cancel()`.

Actually, it turns out I said more than I was planning to say, but I guess it can't hurt!

**Comment by aaugustin** [comment]
> I have a connection handler on the server-side that fires off a number of tasks associated with the connection over the course of the connection's lifetime -- using asyncio.ensure_future(). If one of those tasks encounters an exception, I want the task to be able to close the associated connection (without having to block or create yet another task), which should in turn cause the cancellation / clean-up of the other tasks. When the task initiates the close, this should cause the parent handler to clean-up the tasks it spawned.

I would implement this explicitly with:

```python
def handler(ws, path):
    tasks = [
        asyncio.ensure_future(...),
        asyncio.ensure_future(...),
       asyncio.ensure_future(...),
    ]
    done, pending = await asyncio.wait(tasks, return_when=asyncio.FIRST_EXCEPTION)
    for task in pending:
        task.cancel()
```

**Comment by RemiCardona** [comment]
So I bumped into this using a simple `nc` / `openssl s_client`. @cjerdonek's #310 rebased on top of master does indeed fix the call to `.close()` and prevents the `TypeError`. So 👍 from me.

**Comment by aaugustin** [comment]
Coming back to this issue today, I have a few ideas to move it forwards.

[asyncio.Server.close](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server.close) — which is called by `WebSocketServer.close()` — prevents the server from accepting new connections but leaves existing connections open.

So, once no new connections are coming in, `WebSocketServer.wait_closed()` has two jobs:

1. to terminate all asyncio tasks handling connections, 
2. to terminate the corresponding TCP / TLS connections.

Once that's done, the Python process can terminate cleanly (without pending asyncio tasks) and immediately (without waiting for network connections to terminate — I'm not familiar with that part).

I would find it weird if a connection handler was still running or a TCP connection was still established after `wait_closed` returns. Also being able to terminate a server completely within a fixed timeout is a good property for restarting it in production.

That said, asyncio.Server doesn't appear to track connections after they're established. As far as I can tell from reading the source code, [asyncio.Server.wait_closed](https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.Server.wait_closed) returns when the server stops accepting new connections, not when it's finished dealing with open connections. It doesn't have the properties I described above.

I just discovered that behavior and it's a surprise to me. Should we stick with the behavior we're currently trying to implement — cleaning up everything before wait_closed returns? Or follow asyncio's example and let users sort it out?

For now I'm sticking to my original intention but interested in hearing other opinions :-)

----

This behavior means we should `yield from self.server.wait_closed()` _before_ running our own cleanup code, not _after_ like we currently do.

With the current implementation, new connections can still come in while `WebSocketServer.wait_closed()` is running, which is a bug.

----

Coming back to the two jobs of `WebSocketServer.wait_closed()`, there are two lifetimes to consider for each connection, respectively:

1. the lifetime of `handler_task`; active protocols are tracked in `WebSocketServer.websockets` through the `register` and `unregister` methods;
2. the lifetime of TCP connections; I thought `asyncio.Server` would deal with them for us in `Server.wait_closed`, but in fact it doesn't.

These lifetimes start simultaneously in `connection_made` where `WebSocketServerProtocol` instances register themselves in `server.websockets` and start their `handler_task`. This is reliable.

Currently they unregister themselves when `handler_task` terminates.  When everything goes well, `handler_task` executes `yield from self.close()` successfully and the TCP connection is closed.

There's a lot of error handling code in `WebSocketServerProtocol.handler` to ensure that the TCP connection is closed. I'm not confident that it works.

If, by reasoning on the logic of that code, we can prove that the TCP connection is closed before the corresponding `handler_task` terminates (even when there are cancellations, etc.), that's a very good result:

- all we need is to wait for the `handler_task` of all connections;
- and we already have that list in `WebSocketServer.websockets`!

We also need to make sure that `transfer_data_task` and `close_connection_task` are done if they were started. In practice, making sure that `close_connection_task` is done suffices because:

- it waits for `transfer_data_task` to terminate
- it is never exits with an exception and it is never cancelled

"The TCP connection is closed" and "`close_connection_task` is done" are synonyms when `close_connection_task` has started. When it hasn't, well... it's complicated, because the opening handshake may complete and `close_connection_task` may start at anytime.

Closing at TCP level any connection that hasn't completed the opening handshake at the moment the server shutdown is initiated might work.

----

To sum up:

1. move `yield from self.server.wait_closed()` to the beginning of `WebSocketServer.wait_closed`
2. apply the change shown in #310 — wait only for `handler_task`
3. tweak `WebSocketServerProtocol.handler` to guarantee that (a) the TCP connection is closed (b) if `close_connection_start` was started, it's done

----

Regarding the latest comment by @cjerdonek, `WebSocketCommonProtocol.close()` takes the approach of shutting down the websocket connection. It assumes that the `handler_task` will hit an exception rapidly if it didn't expect that.

You can shut down from the other end by cancelling `handler_task`. The cancellation should be caught, error handling code should kick in and close the websocket connection (in order to bring the guarantees I described above), and only then `handler_task` should exit.

Currently we have three tasks or more (depending on whether the user starts additional tasks from the connection handler). It would be technically feasibly to go back to two tasks, but not to one task, because you need a task to handle control frames in the background while the user's code runs in another task.

My initial design had only one task and relied on the fact that there was one and only one `recv()` running at any time. It was complicated and inconvenient for managing the closing handshake. That's why I introduced a separate task for managing the connection and queues for inter-task communication.

If we do what I described above — and which I believe is already the intent of the current implementation — then `handler_task` is the "root" task. We already document that the server terminates when the connection handler exits. So cancelling the connection handler is a way to terminate the connection.

You can get a reference to `handler_task` with [asyncio.Task.current_task](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.current_task) and use that to cancel it while it's running your code. I'm not sure we want to document that because cancelling `handler_task` when it's running websockets' code (during the opening or closing handshake) could create a mess.

----

All this doesn't help much managing cancellation of parents or children. For that, I think you need the concept of nursery, which is nicely explained here: https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/ but unfortunately not available in asyncio.

To the best of my knowledge, you have to handle all that manually, and it's a major pain.

----

Sorry for the long comment...

**Comment by aaugustin** [comment]
While the TypeError shouldn't happen anymore, I'd like to look into this further.

**Comment by aaugustin** [comment]
Closing in favor of #394 where I summarized this discussion (where my very long comments were getting difficult to follow...)
